### PR TITLE
[PVR][guiinfo][skin] PVR Addon API 5.8.0: Remove unused/deprecated API functions

### PR DESCRIPTION
**Pull request by ksooo**
+ Removed unused API functions `MoveChannel`, `PositionLiveStream`, `PositionRecordedStream`
==================================================================
Todo for addon maintainers: Remove implementation of aforementioned functions from your addons. It's dead code now.

+ Removed depercated API functions `GetPlayingTime`, `GetBufferTimeStart`, `GetBufferTimeEnd`
===================================================================
Todo for addon maintainers: Remove implementation of aforementioned functions from your addons. It's dead code now.  If your addon supports timeshifting, you MUST now implement the API function`GetStreeamTimes`which was introduced with API 5.5.0 some time ago. Otherwise timeshift UI labels and navigating ts buffer (seek, skip, ...) will not work any longer. Even if your addon does not support timeshifting, you SHOULD implement `GetStreamTimes`. Ask @FernetMenta for deatils if you have questions.

+ Addon member `iSubChanneNumber`to struct `PVR_CHANNEL_GROUP_MEMBER`
============================================================
Todo for addon maintainers: This was missing and is needed for proper subchannel number support (ATSC).

+ guiinfo label changes:
================
Renamed labels: 
PVR.Time 	=> PVR.EpgEventElapsedTime
PVR.Duration 	=> PVR.EpgEventDuration
PVR.Progress 	=> PVR.EpgEventProgress
New labels:
PVR.EpgEventRemainingTime
PVR.EpgEventFinishTime

Reason for the changes is new functionality, cleanup of naming and most important, no more hacking epg event times into player.* labels implementation.

@ronie can you please review skin changes and announce the label changes?

Changes are runtime-tested by me and @FernetMenta.

@rbuehlma @janbar @manuelm @dvblogic @margro @fuzzard @sub3 @julianscheel @krustyreturns fyi

### CODE REVIEW

File: addons/skin.estuary/xml/Variables.xml [hunk 1]
```diff
@@ -200,6 +200,7 @@
 	<variable name="SeekTimeLabelVar">
 		<value condition="!String.IsEmpty(Player.SeekNumeric)">[B]$INFO[Player.SeekNumeric(hh:mm:ss)][/B]</value>
 		<value condition="Player.Seeking">$INFO[Player.SeekTime]$INFO[Player.Duration, / ]</value>
+		<value condition="VideoPlayer.HasEpg">$INFO[PVR.Time]$INFO[PVR.Duration, / ]</value>
```
Comment (ronie) [inline addons/skin.estuary/xml/Variables.xml]: ksooo has just asked me to inform you these infolabels are going to be removed. ;-p

**Review by ronie — COMMENTED** [review]
looks good to me.

File: addons/skin.estuary/xml/Variables.xml (hunk 1 above)
Comment (ksooo) [reply addons/skin.estuary/xml/Variables.xml]: Yep.

**Review by FernetMenta — APPROVED** [review]
works also with vnsi

**Comment by ksooo** [comment]
Okay, next is that I will open PRs for the official pvr addons. Please note that I will not be able to implement `GetStreamTimes`for the addons. The maintainers of the addons supporting timeshift (afaik this is pvr.dvblink, pvr.dvbviewer, pvr.mythtv, pvr.vbox, pvr.wmc) MUST do this on their own. I will only remove the obsolete API functions, and possibly I will add support for `PVR_CHANNEL_GROUP_MEMBER::iSubChannelNumber` where it is obvious (because addon's `GetChannel` implementation already fills `PVR_CHANNEL::iSubChannelNumber`).

**Comment by rbuehlma** [comment]
For pvr.zattoo and pvr.teleboy the stream handling is done by inputstream.adaptive. Therefore, I expect that these PVR addons do not need to implement GetStreamTimes. Is that correct?

**Comment by ksooo** [comment]
@rbuehlma yes, that's correct. Only add-ons with own demuxer / own stream implementation need to implement GetStreamTimes

**Comment by rbuehlma** [comment]
I will do the cleanup in pvr.teleboy and pvr.zattoo. Not sure if the are considered  "official pvr addons" anyway ;)

**Comment by dvblogic** [comment]
Just want to acknowledge that I have seen the post. Will do the cleanup and look into new functionality in the new year.
Happy holidays, everyone!

**Comment by krustyreturns** [comment]
I have seen the changes too.  I also won't be able to address this until after the holidays, but I appreciate being included.  Happy holidays and thanks for your great work.

**Comment by sub3** [comment]
I've seen the changes too. Thanks for the heads up.

**Comment by ksooo** [comment]
> I will do the cleanup in pvr.teleboy and pvr.zattoo. Not sure if the are considered "official pvr addons" anyway

 They are official if they are included in https://github.com/xbmc/repo-binary-addons

**Comment by manuelm** [comment]
>  If your addon supports timeshifting, you MUST now implement the API function `GetStreamTimes` which was introduced with API 5.5.0 some time ago. Otherwise timeshift UI labels and navigating ts buffer (seek, skip, ...) will not work any longer. Even if your addon does not support timeshifting, you SHOULD implement `GetStreamTimes`.

Implementing `GetStreamTimes` requires returning PTS values. So does timeshifting now requires demuxing the stream? What's the best/recommended way to get those values?

**Comment by ksooo** [comment]
> So does timeshifting now requires demuxing the stream?

 No.

> What's the best/recommended way to get those values?

 This is how I did it for pvr.hts : https://github.com/kodi-pvr/pvr.hts/pull/346/files#diff-fc18272d30993f8ff9965f332263834cR310

**Comment by ksooo** [comment]
@FernetMenta @janbar @julianscheel please review and merge the PR I created for your addon.

@rbuehlma @manuelm @dvblogic @sub3 @krustyreturns please open PRs for your addons (you said you wanted to do this yourself).

**Comment by rbuehlma** [comment]
The "cleanup" is already merged for pvr.zattoo and pvr.teleboy

**Comment by janbar** [comment]
The PR has been merged into pvr.mythtv. Thanks.

**Comment by manuelm** [comment]
@ksooo I tried to implement this, however I've troubles with the time information in the playback overlay being wrong. The calculated play time (`GetPlayTime()`) is completely off.

e.g. Live TV with timeshift disabled, EPG entry started at 1514984400 (14:00 CET), `startTime` set to the current timestamp, the rest set to `0` I get a current play time of 14926 seconds.

Switching to another show on another channel, which started at 1514984100 (13:55 CET) shows an play time of 68969 seconds, which means the player thinks I'm almost a day in the future.

I've tried various combinations without any success.

With returning `PVR_ERROR_NOT_IMPLEMENTED` at first the current play time seems to be correct. However after a second or two the play time seems to get doubled endlessly.

**Comment by ksooo** [comment]
@manuelm please ask @FernetMenta for help how to implement GetStreamTimes correctly. Have you had a look at what I did for pvr.hts?

**Comment by manuelm** [comment]
Yeah. I've based my initial implementation on this, however as it didn't work out I've tried __a lot__ of other combinations but was unable to find a pattern.

**Comment by ksooo** [comment]
For pvr.hts:

 times->startTime is always the time_t the playback of the channel started. I never reset it when playing show changes.

 times->ptsStart is always 0 (the base of relative time calculations for ptsBegin and eptsEnd)

  times->ptsBegin is the delta "start of ts buffer minus start of playback" in seconds * DVD_TIME_BASE

 times->ptsEnd is the delta "end of ts buffer minus start of playback" in seconds * DVD_TIME_BASE

**Comment by margro** [comment]
I'm also having problems with the GetStreamTimes() functionality. After this PR, seeking does not work anymore for recordings and Live TV (MediaPortal PVR addon) fails after 1 second of playback. I tried to implement the GetStreamTimes() functionality for recordings first based on what @FernetMenta did for pvr.vdr.vnsi and I cannot get a working seek functionality at all.
My relevant commit: https://github.com/margro/pvr.mediaportal.tvserver/commit/3a19dc904ec3413640ad40dc2175a8e8b5dc8a7a
Basically, I'm only setting ptsEnd to the length of the recording (in seconds) as documented in the header and I'm not seeing the end time of the recording in Kodi and seeking also does not work.

I also don't understand why we suddenly have to provide pts time information here. I'm reading a file and I don't have knowledge at the addon side about the streams inside the file. I don't have demuxer info avalable. I only have the total file size and the actual position for the timeshift buffer. For recordings that are finished, I know the total duration but not for the timeshift and in progress recordings.

**Comment by ksooo** [comment]
> I'm reading a file and I don't have knowledge at the addon side about the streams inside the file.

 You don't need to know what is inside the file. pts is nothing else than a relative time stamp - relative to a 'start' value. All you need to know how large your ts buffer is in seconds.

**Comment by margro** [comment]
@ksooo and that is the info that the pvr.mediaportal.tvserver and pvr.argustv for example do not have...
There is no backend API info that can provide me with this information. I do have the file size but not the time.
The code from pvr.vdr.vnsi that I've also implemented for MediaPortal is: https://github.com/FernetMenta/pvr.vdr.vnsi/blob/master/src/VNSIRecording.cpp#L127

**Comment by manuelm** [comment]
```
times->startTime is always the time_t the playback of the channel started. I never reset it when playing show changes.
```

Same here (https://github.com/manuelm/pvr.dvbviewer/blob/47973567fd49115974c799e2f148f0bfa7b314e0/src/client.cpp#L545). I also don't have any information about the current show, so I cannot change it anyway.

```
times->ptsStart is always 0 (the base of relative time calculations for ptsBegin and ptsEnd)
times->ptsBegin is the delta "start of ts buffer minus start of playback" in seconds * DVD_TIME_BASE
times->ptsEnd is the delta "end of ts buffer minus start of playback" in seconds * DVD_TIME_BASE
```

For Live TV both `ptsBegin` and `ptsEnd` should be set to 0 according to the documentation. However changing begin or end neither influences the calculated play time.

**Comment by ksooo** [comment]
> For Live TV both ptsBegin and ptsEnd should be set to 0 according to the documentation.

 No. That's for recordings. :-) 

https://github.com/xbmc/xbmc/blob/master/xbmc/addons/kodi-addon-dev-kit/include/kodi/xbmc_pvr_types.h#L599

**Comment by ksooo** [comment]
Sory guys, I really want to @FernetMenta to jump in at this point as he came up with this API.

**Comment by manuelm** [comment]
@ksooo I'm not talking about timeshifting. Just live tv.

@margro For recordings I'm just returning PVR_ERROR_NOT_IMPLEMENTED. Everything else makes the times go wrong for me.

**Comment by manuelm** [comment]
@ksooo I'm wondering if I'm seeing some bug you don't see. e.g if I return `PVR_ERROR_NOT_IMPLEMENTED` without any values I'm getting this debug output (added to `CPVRGUIInfo::UpdateTimeshift`):
```
16:55:36.120 T:140707960452864   DEBUG: UpdateTimeshift: m_bIsTimeshifting=0 m_iStartTime=1514994936 iMinTime=0 iMaxTime=0 iPlayTime=0
16:55:37.121 T:140707960452864   DEBUG: UpdateTimeshift: m_bIsTimeshifting=0 m_iStartTime=1514994936 iMinTime=0 iMaxTime=0 iPlayTime=0
16:55:38.122 T:140707960452864   DEBUG: UpdateTimeshift: m_bIsTimeshifting=0 m_iStartTime=1514994936 iMinTime=0 iMaxTime=0 iPlayTime=0
16:55:39.122 T:140707960452864   DEBUG: UpdateTimeshift: m_bIsTimeshifting=0 m_iStartTime=1514994936 iMinTime=0 iMaxTime=0 iPlayTime=1835
16:55:40.123 T:140707960452864   DEBUG: UpdateTimeshift: m_bIsTimeshifting=0 m_iStartTime=1514994936 iMinTime=0 iMaxTime=0 iPlayTime=3670
16:55:41.124 T:140707960452864   DEBUG: UpdateTimeshift: m_bIsTimeshifting=0 m_iStartTime=1514994936 iMinTime=0 iMaxTime=0 iPlayTime=5504
16:55:42.125 T:140707960452864   DEBUG: UpdateTimeshift: m_bIsTimeshifting=0 m_iStartTime=1514994936 iMinTime=0 iMaxTime=0 iPlayTime=7339
16:55:43.129 T:140707960452864   DEBUG: UpdateTimeshift: m_bIsTimeshifting=0 m_iStartTime=1514994936 iMinTime=0 iMaxTime=0 iPlayTime=9173
16:55:44.131 T:140707960452864   DEBUG: UpdateTimeshift: m_bIsTimeshifting=0 m_iStartTime=1514994936 iMinTime=0 iMaxTime=0 iPlayTime=11007
```

The calculated play time is jumping within seconds.

**Comment by ksooo** [comment]
>The calculated play time is jumping within seconds.

 It's increasing because you are playng the channel, I guess?

**Comment by margro** [comment]
A quick question. Are the pts values in seconds or in microseconds?
De xbmc_pvr_types.h header says: "Value is seconds" for ptsBegin and ptsEnd, where @ksooo multiplies the seconds with DVD_TIME_BASE which is 1000000.

**Comment by manuelm** [comment]
But it's increasing way to fast :-) 11007 "play time"-seconds within 8 realtime seconds.

**Comment by ksooo** [comment]
> But it's increasing way to fast :-) 11007 "play time"-seconds within 8 realtime seconds.

 No idea.

**Comment by manuelm** [comment]
@margro Actually doesn't matter as long as all values have the same unit. VideoPlayer.cpp is only using differences (end - start, begin - start). I used seconds and the timeshift information was correct.

**Comment by ksooo** [comment]
Right. I think I will remove this (leftover) from pvr.hts (and documentation should be improved)

**Comment by ksooo** [comment]
> Right. I think I will remove this (leftover) from pvr.hts (and documentation should be improved)

 I have to correct myself. Currently, it must be "seconds times DVD_TIME_BASE". Otherwise it semms not to work. Maybe this is a Kodi bug, don't know. @FernetMenta should know.

**Comment by ksooo** [comment]
There is a line in VideoPlayer.cpp  (line 4609 in current master): 

`state.time = (m_clock.GetClock(false) - times.ptsStart) * 1000 / DVD_TIME_BASE;`.

This implies that `times.ptsStart` must have the same unit as what `m_clock.GetClock` returns. `times.ptsStart`directly comes from the addon, `m_clock.GetClock` is Kodi internal.

**Comment by margro** [comment]
@ksooo and @manuelm I can confirm that these pts values in "seconds times DVD_TIME_BASE" (=micro seconds) seems to work currently. I can seek my recordings again. Next challenge is to fix the Live TV playback...

**Comment by FernetMenta** [comment]
@ksooo 
pts is in micros. comment seems to be wrong:
https://github.com/xbmc/xbmc/blob/master/xbmc/addons/kodi-addon-dev-kit/include/kodi/xbmc_pvr_types.h#L599

**Comment by ksooo** [comment]
yeah. I will fix the documentation.

**Comment by manuelm** [comment]
@FernetMenta Any idea about my issue with the play time? e.g.

xxx

shows a play time of 22h56mins. This is live TV with timeshift disabled. So my `GetStreamTimes` only sets startTime to the start of the playback, all other members are set to `0`.

**Comment by rbuehlma** [comment]
To which value do you set "startTime"? Kodi does not handle values that exceed MAX_INT32. In my case (using inputstream.adaptive for stream handling), PTS is the unix time stamp which does not fit in INT32 as micro secs. The behaviour is about the same.

**Comment by manuelm** [comment]
`time(NULL)` which get's assigned on `OpenLiveStream`.

**Comment by manuelm** [comment]
I traced the `m_clock.GetClock(false)` calls in VideoPlayer.cpp and it seems on playback the internal clock quickly rises and so does my play time:
```
19:11:00.215 T:140552091711232   DEBUG: VideoPlayer: UpdatePlayState: clock=8.335000 state.time=0.007754
19:11:00.427 T:140552091711232   DEBUG: VideoPlayer: UpdatePlayState: clock=200881.475000 state.time=200.881151
19:11:00.627 T:140552091711232   DEBUG: VideoPlayer: UpdatePlayState: clock=391032.245000 state.time=391.032002
19:11:00.829 T:140552091711232   DEBUG: VideoPlayer: UpdatePlayState: clock=583094.114000 state.time=583.093670
19:11:01.152 T:140552091711232   DEBUG: VideoPlayer: UpdatePlayState: clock=889633.832000 state.time=889.633353
19:11:01.332 T:140552091711232   DEBUG: VideoPlayer: UpdatePlayState: clock=83874777349.492996 state.time=83874777.349239
19:11:01.533 T:140552091711232   DEBUG: VideoPlayer: UpdatePlayState: clock=83874967748.334000 state.time=83874967.747369
19:11:01.802 T:140552091711232   DEBUG: VideoPlayer: UpdatePlayState: clock=83875223443.505005 state.time=83875223.443093
```
No idea if this is normal behavior or what I should do about this.

**Comment by manuelm** [comment]
Hardcoding `m_iTimeshiftPlayTime` in `xbmc/pvr/PVRGUIInfo.cpp` to `std::time(nullptr)` fixes the play time for me. Ofc this is no viable workaround. For whatever reason `iPlayTime` is not accurate with the dvbviewer backend. I'm also getting first reports of nightly users.

@FernetMenta I could really need your help here.

**Comment by FernetMenta** [comment]
@manuelm 
> This is live TV with timeshifting disabled. So my GetStreamTimes only sets startTime to the start of the playback, all other members are set to 0.

This makes no sense. If you don't want to set a start/end time, you should not return NO_ERROR on this method.

**Comment by FernetMenta** [comment]
> Implementing GetStreamTimes requires returning PTS values. So does timeshifting now requires demuxing the stream? What's the best/recommended way to get those values?

ptsBegin and ptsEnd, in units of microseconds, use ptsStart as reference. Assume the addon has no demuxer. That means that ffmpeg demuxer sets pts at start to zero. Now assume that you tune to a channel with timeshift active. You would set startTime to some time in UTC, startPts to 0, and ptsBegin/End to values indicating begin/start of the timeshift buffer. This would be most likely 0/0 adn the very begining. ptsEnd will increase in the same speed as the stream fills the buffer.

**Comment by manuelm** [comment]
> This makes no sense. If you don't want to set a start/end time, you should not return NO_ERROR on this method.

This happens if I return `PVR_ERROR_NOT_IMPLEMENTED`: https://manuel.mausz.at/kodi_playtime.ogv

**Comment by manuelm** [comment]
Example stream from the backend: https://manuel.mausz.at/kodi_dvbstream.ts

**Comment by FernetMenta** [comment]
@ksooo if GetStreamTimes does not return NO_ERROR, the fallback path IDisplayTime is taken. Looks like it is broken.

**Comment by ksooo** [comment]
Okay, I will take a look, but what is not clear to me is when it is legit to return NOT_IMPLEMENTED. Could you please elaborate on this?

**Comment by manuelm** [comment]
Also I see similar behavior with timeshifting enabled. The playtime immediately jumps to a high (+random) value. Timeshifting is local, so always starts at playback (or pause) and runs until stream stop. Imho this has something to do with the PTS values inside the stream and/or Kodi's internal clock: https://github.com/xbmc/xbmc/pull/13228#issuecomment-355087785

**Comment by FernetMenta** [comment]
addons that have NOT implemented it return NOT_IMPLEMENTED. This path seems to be broken.

**Comment by FernetMenta** [comment]
@manuelm if an addon supports timeshift it MUST implement GetStreamTImes properly

**Comment by ksooo** [comment]
> This path seems to be broken.

 I get that and I will look into it.

**Comment by manuelm** [comment]
> @manuelm if an addon supports timeshift it MUST implement GetStreamTImes properly

I *did* implement it properly. `startTime` gets set to time(NULL) on timeshift start. `ptsStart` + `ptsBegin` is set to 0. No other variable influences the play time. I'm doing the same as other PVRs and I still get weird behavior.

**Comment by ksooo** [comment]
@manuelm please give us some time to check whether this is a core bug, okay?

**Comment by FernetMenta** [comment]
@manuelm please point me to your code

**Comment by manuelm** [comment]
Sure. I've pushed the 5.8.0 adaption to https://github.com/manuelm/pvr.dvbviewer/tree/pvrtimes yesterday. I do know that in the timeshifting case `ptsEnd` has the wrong unit (s vs ms), however that doesn't change the play time behaviour.

I'm off skiing for two days now. I'll try to create a demo pvr on sunday playing the above TS file for you.

**Comment by dvblogic** [comment]
@ksooo @FernetMenta 
I thnk I might also use some help from the guru's. The timeshifted stream handling puzzle does not fit together for me. 
With the removal of PositionLiveStream, how does the player know its current position inside the buffer? 
GetStreamTimes reports only start and end timestamps of the buffer.
LengthLiveStream (what is it used for at all?) is only called once at the beginning of playback.
SeekLiveStream function operates in bytes, but current size of timeshift buffer is not asked anywhere, except for LengthLiveStream at the start of the streaming.

And on top of that there is an additional function, who's meaning is also not clear to me - IsRealTimeStream. What should it return in case of timeshift enabled? true, when close to the real-time end of the buffer and false otherwise?

Thanks.

**Comment by FernetMenta** [comment]
> With the removal of PositionLiveStream, how does the player know its current position inside the buffer?

videoplayers knows by pts/dts of the stream.

> And on top of that there is an additional function, who's meaning is also not clear to me - IsRealTimeStream. What should it return in case of timeshift enabled? true, when close to the real-time end of the buffer and false otherwise?

That's what vnsi does.

**Comment by dvblogic** [comment]
Thanks a lot.
> videoplayers knows by pts/dts of the stream.
This would probably explain the displayed time jumps that others also observe as pts values in the live stream do not start with 0.

How about the SeekLiveStream function that seeks in bytes? How does it know the buffer size? And what the LengthLiveStream function should return?

**Comment by FernetMenta** [comment]
> This would probably explain the displayed time jumps that others also observe as pts values in the live stream do not start with 0.

For addons with no demuxer pts at start is 0 because internal demuxer sets an offset to make it 0.

SeekLiveStream, SeekLiveStream: dead code

**Comment by FernetMenta** [comment]
Read and SeekLiveStream a anlogue read/seek of ordinary file functions. Demuxer might use SeekLiveStream more than once in order to find desired timestamp.
API for inputstream also provides IPosTime interface that allows an addon to seek directly to a given time. This will be available to PVR once the streaming API got split from the rest of PVR.

**Comment by bernhardberger** [comment]
@manuelm any news on this front?

**Comment by dvblogic** [comment]
I would like to check if any progress was made by anyone on the issue where Timeshift time jumps to some high bogus value after about 10 seconds of starting playback?
I checked it against the latest Leia development build and it still happens.

Also, have not seen any commits yet for the new PVR API by any of the addon developers (except for hts), so cannot check my implementation against any other without demux.

Any assistance is welcome.
Thanks.

**Comment by manuelm** [comment]
@dvblogic I guess you are talking about the issue I described here. In https://github.com/manuelm/pvr.dvbviewer/issues/21#issuecomment-363758741 @FernetMenta confirmed that this is an issue in Kodi / ffmpeg. But no solution yet.

**Comment by dvblogic** [comment]
@manuelm Thanks. for the information. I think we would need to wait a bit.

**Pull request merged**